test: set up authentication

JeffreytheCoder opened the pull request:
# What this PR does
- Set up authentication for E2E tests. Simulate login flow using user credentials stored in environment variables. Run once before all test suites by saving auth file to storage state
- Modified Playwright configuration to correctly run the tests
- Added an example test to see if EC renders
- Wrote README on how to set up tests

## Acceptance Criteria fulfillment

- [ ] `login()`

Fixes #546 
## Video/Screenshots
![test-auth](https://github.com/RocketChat/EmbeddedChat/assets/35394596/f7174bd2-ea60-444d-8849-1e16468b5adf)

JeffreytheCoder commented:
The test workflow works when `baseURL` is `http://127.0.0.1:5173` in `playwright.config.ts`, but I changed it to `http://localhost:6006` to run tests against EC in storybook, which makes the workflow failed. @sidmohanty11 Any idea why this happened?

Spiral-Memory commented:
Hey @JeffreytheCoder , I have some questions in mind. I don't have much understanding in setting up these tests, but I want to learn more about it so that I can also contribute by writing e2e tests, as the central issue is huge. I was going through your changes, and I would like to ask a few questions. It would be great if you could spare some of your time answering them and help me understand.

Spiral-Memory commented:
What each of these is doing and why did you keep 2 different ports for this? Earlier both of them were the same. After changing it to localhost:6000, I can see that Playwright tests are failing in your PR, and it says that it is unable to connect to that, and the connection is refused. However, before, it used to connect and have a check on basic things like whether the chat header is present or not, and it obviously happens if it is able to see the UI. I might be asking silly question, I am not exactly how it works properly, could you please help me understand it ?

<img width="406" alt="image" src="https://github.com/RocketChat/EmbeddedChat/assets/78961432/3b6540cd-83e7-4062-9ca0-65e8dceac353">
<img width="458" alt="Screenshot 2024-04-14 104747" src="https://github.com/RocketChat/EmbeddedChat/assets/78961432/e9714740-11b8-4201-bf84-e6f82e22200a">

Spiral-Memory commented:
I would like to know why you removed the host. Without a host, how will it connect to the Rocket.chat server? I believe chat.avitechlab.com was a hosted Rocket.Chat server on which api call could be made and tested on. Could you please help me understand this change?

<img width="627" alt="image" src="https://github.com/RocketChat/EmbeddedChat/assets/78961432/9c845316-917e-4605-8a68-16ba295c71f4">

JeffreytheCoder commented:
Hey @Spiral-Memory I was confused with whether the tests should run in development environment or in a dedicated testing environment in CI. I changed host to storybook and added environment variables so that each EC dev can run tests using his/her own server. However, if we only want to run tests during CI, I'll change it back to using `chat.avitechlab.com`.

JeffreytheCoder commented:
Before this PR, EC was set to anonymous mode for CI tests. I was thinking we can just discard the authentication setup, but then I saw EC currently only allows anonymous read, not anonymous write. If this is a feature, not a bug, we then need a test account of the hosted server to sign in and test the messaging features.

Spiral-Memory commented:
> Hey @Spiral-Memory, I was confused about whether the tests should run in the development environment or in a dedicated testing environment in CI. I changed the host to Storybook and added environment variables so that each EC dev can run tests using their own server. However, if we only want to run tests during CI, I'll change it back to using `chat.avitechlab.com`.

In my opinion, tests should run on CI to achieve an automated workflow. Additionally, the environment variables for login are fine; the repository owner can add those environment variables to enable automated login on `chat.avitechlab.com`.

Also, we have two modes in RC. One is anonymous mode, which allows us to read messages if the workspace owner has enabled it. So, there's no need to remove that. Let's check for the anonymous mode first. Once the test passes, we can proceed to login and test the other features which require write access accordingly.

Spiral-Memory commented:
> What each of these is doing and why did you keep 2 different ports for this? Earlier both of them were the same. After changing it to localhost:6000 and localhost:5173, I can see that Playwright tests are failing in your PR, and it says that it is unable to connect to that, and the connection is refused maybe because server is started on different port and you are trying to connect to a different port. However, before, it used to connect and have a check on basic things like whether the chat header is present or not, and it obviously happens if it is able to see the UI. I might be asking silly question, I am not exactly how it works properly, could you please help me understand it ?
> 
> <img alt="image" width="406" src="https://private-user-images.githubusercontent.com/78961432/322248817-3b6540cd-83e7-4062-9ca0-65e8dceac353.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMwNzUzODAsIm5iZiI6MTcxMzA3NTA4MCwicGF0aCI6Ii83ODk2MTQzMi8zMjIyNDg4MTctM2I2NTQwY2QtODNlNy00MDYyLTljYTAtNjVlOGRjZWFjMzUzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE0VDA2MTEyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMxNTgyMGJmZjVjMDQyZjIyYmNmM2MxZWI1NmFiMjQ5YTk5YTY1OTBmNmY3MmE1YzI4ODcwYTFiOWU1OWNjM2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.nNxjMM3YObbHvD-UOFJFnWs9klHBg00XQVUkpo5GVes"> <img alt="Screenshot 2024-04-14 104747" width="458" src="https://private-user-images.githubusercontent.com/78961432/322248833-e9714740-11b8-4201-bf84-e6f82e22200a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMwNzUzODAsIm5iZiI6MTcxMzA3NTA4MCwicGF0aCI6Ii83ODk2MTQzMi8zMjIyNDg4MzMtZTk3MTQ3NDAtMTFiOC00MjAxLWJmODQtZTZmODJlMjIyMDBhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE0VDA2MTEyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZlM2JmOGM1ZjlkYzc5NDViNGM4YjJiYjBmYjI4ODk0ZDM1N2Q2NmE5NWI3MGM1YThiZTRkMDVhYzY3MjUzNzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.VcX13-TN6hvFn7sPRaoW5z7xlB8ozSkvKyQOMZcF3UE"> <img alt="image" width="601" src="https://private-user-images.githubusercontent.com/78961432/322249136-dd1a0ddd-00ad-408a-a3d7-38e7b5f9907b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMwNzUzODAsIm5iZiI6MTcxMzA3NTA4MCwicGF0aCI6Ii83ODk2MTQzMi8zMjIyNDkxMzYtZGQxYTBkZGQtMDBhZC00MDhhLWEzZDctMzhlN2I1Zjk5MDdiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE0VDA2MTEyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgwNDQ5MzQ2YjBkMDBjYWRkYjY5MWYyOWMwMDc2NmM3MmI4YzI0NWE1YWEzMGU5MWM1NGJlYWIyOWMxNjQ2NWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.7Fu-ZwMU5Z5LTttXxbK6UNbKdvqJqb0o3nUiBI1nrZQ">

Also, once we refer to this, I think one is to start the EC, and the other part will go to that URL and test. So, keep it consistent, and it should be the same to successfully test things. Again, I'm not very sure how testing works, but this is what makes sense to me. Let's give it a try.

JeffreytheCoder commented:
Hey @abhinavkrin @sidmohanty11 do you guys have the credentials to the test server chat.avitechlab.com?

Spiral-Memory commented:
> Hey @abhinavkrin @sidmohanty11 do you guys have the credentials to the test server chat.avitechlab.com?

Obviously they will have ig 😂, for testing, btw you can create your own account on that Rocket.Chat server for now. I just created one for me. (but not with admin access obviously)

JeffreytheCoder commented:
> > Hey @abhinavkrin @sidmohanty11 do you guys have the credentials to the test server chat.avitechlab.com?
> 
> Obviously they will have ig 😂, for testing, btw you can create your own account on that (avitech) Rocket.Chat server for now. I just created one for me. (but not with admin access obviously)

Ohh that's great. I thought we need to share a test account so that's why I asked haha

Spiral-Memory commented:
> > > Hey @abhinavkrin @sidmohanty11 do you guys have the credentials to the test server chat.avitechlab.com?
> > 
> > 
> > Obviously they will have ig 😂, for testing, btw you can create your own account on that (avitech) Rocket.Chat server for now. I just created one for me. (but not with admin access obviously)
> 
> Ohh that's great. I thought we need to share a test account so that's why I asked haha

It depends on Abhinav.

Also, we might need to test some features with admin access, so I'm a little unsure if Abhinav will provide a test user with admin access. Haha... Maybe we can test non-admin related features with our own credentials on that server or using a common shared test user (without admin access / with admin access, depending on Abhinav's choice). But I think he is quite busy these days, so for now let's use our credentials to conduct testing on the local system (either with our local server (if admin access is required) or with the Avitechlab server).

JeffreytheCoder commented:
Anyway we need a shared account for CI to use. I created a tester account and we can change to an admin account later
<img width="565" alt="Screenshot 2024-04-13 at 11 41 57 PM" src="https://github.com/RocketChat/EmbeddedChat/assets/35394596/0c8e3fcd-e661-43ef-b85f-0b082121f0f9">

Spiral-Memory commented:
Yup ! Great thanks for sharing

Spiral-Memory commented:
Anyway, let's start working on it.. once the login is complete and once you make the required changes to this PR. Let me know, we will test together !!

JeffreytheCoder commented:
Added a new commit and the tests now work in CI :)

Spiral-Memory commented:
@sidmohanty11 Could you once add credentials in GitHub workflow directly so that it can be fetched from. env rather than hardcoding a test user.

JeffreytheCoder commented:
> > Added a new commit and the tests now work in CI :)
> 
> Great but you have removed some tests, like seeing the messages and stuff from the example.spec.ts, once revert those as well.
> 
> Also, after logging in, we need to check things like whether the input box is now enabled, whether the "Login successful" toast message is rendered, etc., to verify that the login has been successful, in my opinion. **Please mark it as a draft for now, and once it's done, make it ready for review to avoid confusion.**

Let's wait until the meeting to verify these changes
